src/modules/esl-share/core/esl-share-config.ts

```diff
@@ -33,120 +34,132 @@ export interface ESLShareGroupConfig {
+   * Updates the configuration with promise resolved to {@link ESLShareConfigInit} or promise provider function.
    * Every button and group specified in the new config will be added to the current configuration.
-   * @returns ESLShareConfig instance
+   * @returns Promise<ESLShareConfig> instance
    */
   public static set(
-    provider?: ESLShareConfigProviderType | Promise<Partial<ESLShareConfig>> | Partial<ESLShareConfig>
-  ): ESLShareConfig {
-    const {instance} = ESLShareConfig;
+    provider: Promise<Partial<ESLShareConfig>> | PropertyProvider<Promise<ESLShareConfigInit>>
+  ): Promise<ESLShareConfig>;
+  public static set(provider: any): ESLShareConfig | Promise<ESLShareConfig> {
     if (typeof provider === 'function') return this.set(provider());
-    if (typeof provider === 'object' && provider instanceof Promise) {
-      provider.then((cfg) => this.set(cfg));
-      return instance;
-    }
-    if (typeof provider === 'object') {
-      if (provider?.groups) provider.groups.forEach(instance.appendGroup, instance);
-      if (provider?.buttons) provider.buttons.forEach(instance.appendButton, instance);
-    }
-    return instance;
+    if (!isObject(provider)) return ESLShareConfig.instance;
+    if (provider instanceof Promise) return provider.then((cfg: ESLShareConfigInit) => this.set(cfg));
+    if (Array.isArray(provider.groups)) ESLShareConfig.instance.append(provider.groups);
+    if (Array.isArray(provider.buttons)) ESLShareConfig.instance.append(provider.buttons);
+    return ESLShareConfig.instance;
+  }
+
+  /** Appends single button or group to current configuration */
+  public static append(cfg: ESLShareButtonConfig | ESLShareGroupConfig | ESLShareButtonConfig[] | ESLShareGroupConfig[]): ESLShareConfig {
+    return ESLShareConfig.instance.append(cfg);
   }
 
-  public readonly buttons: ESLShareButtonConfig[] = [];
-  public readonly groups: ESLShareGroupConfig[] = [];
+  protected readonly _groups: Map<string, string> = new Map();
+  protected readonly _buttons: Map<string, ESLShareButtonConfig> = new Map();
 
   protected constructor() {
     super();
   }
 
+  /** @returns list of all available groups */
+  public get groups(): ESLShareGroupConfig[] {
+    return Array.from(this._groups.entries()).map(([name, list]) => ({name, list}));
+  }
+
+  /** @returns list of all available buttons */
+  public get buttons(): ESLShareButtonConfig[] {
+    return Array.from(this._buttons.values());
+  }
+
+  /** Normalize list string by removing groups and extra whitespaces */
+  public resolve(query: string): string[] {
+    if (query.includes('all')) return Array.from(this._buttons.keys());
+    const groups = new Set<string>(); // Deduplicate groups
+    while (query.includes('group:')) {
+      query = query.replace(/group:(\S+)/gi, (term: string, name: string) => {
+        if (groups.has(name)) return '';
+        groups.add(name);
```

Comment:
Should we be engaged in optimization? It seems to me that this is a content issue and for the component - unnecessary responsibilities.

Reply:
Here, we have a terrible case (infinite loop with super small leak - so the page just freezes) in case users do this. 
And unlike the simple project, we do not have the option to say it's content in case we have a crash.